### PR TITLE
feat: native map/array constructors

### PR DESCRIPTION
**Pull request by agavra**
BREAKING CHANGE: the old syntax for creating arrays using the
builtin AS_ARRAY function will no longer work

### Description 

Follow-up on #4120 - this PR creates maps and arrays natively instead of going through the UDF path. This has the benefit of custom syntax and stricter type checking.

### Testing done 

- QTT
- Unit Tests

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by agavra** [comment]
Thanks for the review @big-andy-coates! I agree with your comments about testing and casting and I'll add those in. 

I don't think we should keep `AS_ARRAY` because anyone that is upgrading to 0.7 will already have lots of breaking changes to make and this one isn't hard to change.

For the array constructor syntax, I think I'd prefer to keep it as is (which is the sql99 standard: http://web.cecs.pdx.edu/~len/sql1999.pdf):

> `<array value list constructor> ::= ARRAY <left bracket or trigraph> <array element list> <right bracket or trigraph>`

A `trigraph` is `??(`, which very few sql implementations seem to support. All examples I've found online of sql systems use the `ARRAY[]` syntax.

**Comment by blueedgenick** [comment]
I'm fully supportive of the square-bracket syntax for array creation - it
is, after all, a language construct and not a function (at least it is now!
) and this matches the systems i've used in the past.

Introducing new syntax for CASTing, a la postgres, seems iffy though
without further discussion.

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java [hunk 1]
```diff
@@ -331,6 +333,83 @@ public Void visitSubscriptExpression(
       return null;
     }
 
+    @Override
+    public Void visitCreateArrayExpression(
+        final CreateArrayExpression exp,
+        final ExpressionTypeContext context
+    ) {
+      if (exp.getValues().isEmpty()) {
+        throw new KsqlException("Cannot extract type from empty array!");
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: https://github.com/confluentinc/ksql/issues/4239 will link to that issue

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java [hunk 2]
```diff
@@ -331,6 +333,83 @@ public Void visitSubscriptExpression(
...
+          .map(val -> {
+            process(val, context);
+            return context.getSqlType();
+          })
+          .filter(Objects::nonNull)
+          .distinct()
+          .collect(Collectors.toList());
+
+      if (sqlTypes.size() == 0) {
+        throw new KsqlException("Cannot extract type from array of nulls!");
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: > Can the user work around this by casting one of the NULLs to a type? Maybe include this in the error message and add a QTT test case.

Yes and done.

> What about for INSERT VALUES where the required array type is known.... can we support all nulls then? How about empty arrays?

This is not a trivial change. I've thought about this in the context of various things (i.e. struct type inference) and it's out of scope for this PR, though technically possible.

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java [hunk 3]
```diff
@@ -331,6 +333,83 @@ public Void visitSubscriptExpression(
...
+      return null;
+    }
+
+    @Override
+    public Void visitCreateMapExpression(
+        final CreateMapExpression exp,
+        final ExpressionTypeContext context
+    ) {
+      if (exp.getMap().isEmpty()) {
+        throw new KsqlException("Cannot extract type from empty map!");
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: BUT eRRoRS Are SO ExCITING!!11!11!11

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java [hunk 4]
```diff
@@ -331,6 +333,83 @@ public Void visitSubscriptExpression(
...
+          .map(val -> {
+            process(val, context);
+            return context.getSqlType();
+          })
+          .distinct()
+          .collect(Collectors.toList());
+
+      if (sqlTypes.size() != 1) {
+        throw new KsqlException(
+            "Invalid map expression! All values must be of the same type. " + exp);
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: Added a better error message, but I don't think we need to tell the user to `CAST` as that may not be the right solution for them.

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java [hunk 5]
```diff
@@ -331,6 +333,83 @@ public Void visitSubscriptExpression(
...
+      }
+
+      if (sqlTypes.get(0) == null) {
+        throw new KsqlException("Maps do not accept null values!");
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: See above

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java [hunk 6]
```diff
@@ -331,6 +334,94 @@ public Void visitSubscriptExpression(
...
+            process(val, context);
+            return context.getSqlType();
+          })
+          .filter(Objects::nonNull)
+          .collect(Collectors.toList());
+
+      if (sqlTypes.size() == 0) {
+        throw new KsqlException("Cannot construct an array with all NULL elements "
+            + "(see https://github.com/confluentinc/ksql/issues/4239). As a workaround, you may "
+            + "cast the NULL value to a desired type.");
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: nit: consider changing 'the NULL' to 'a NULL'.  There could be multiple nulls and only one needs to be none-null

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: Can we link in a github issue to track this second part please?

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: What's the issue with suggesting `CAST` as a possible solution? We're not forcing them to use it...

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java [hunk 7]
```diff
@@ -331,6 +334,94 @@ public Void visitSubscriptExpression(
...
+            process(key, context);
+            return context.getSqlType();
+          })
+          .collect(Collectors.toList());
+
+      if (gkeyTypes.stream().anyMatch(type -> !SqlTypes.STRING.equals(type))) {
+        throw new KsqlException("Only STRING keys are supported in maps but got: " + gkeyTypes);
+      }
+
+      final List<SqlType> sqlTypes = exp.getMap()
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: super nit: can we call this `valueTypes` and the other `keyTypes` to make it clear?

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: Still think the error message could be improve to:

| Can not construct MAP with all NULL values.

And add the trick of using `CAST`.

File: ksql-functional-tests/src/test/resources/query-validation-tests/asarray.json [hunk 8]
```diff
@@ -7,7 +7,7 @@
       "name": "construct a list from two elements",
       "statements": [
         "CREATE STREAM TEST (a INT, b INT) WITH (kafka_topic='test_topic', value_format='JSON');",
-        "CREATE STREAM OUTPUT AS SELECT as_array(a, b, 3) as l FROM TEST;"
+        "CREATE STREAM OUTPUT AS SELECT ARRAY[a, b, 3] as l FROM TEST;"
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/asarray.json]: might want to rename this test file... it's currently `asarray.json` and `AS_ARRAY` no longer exists ;) . How about `arrays.json` :p

File: ksql-functional-tests/src/test/resources/query-validation-tests/asmap.json [hunk 9]
```diff
@@ -3,6 +3,19 @@
     "Tests covering map creation"
   ],
   "tests": [
+    {
+      "name": "create map from named tuples",
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/asmap.json]: likewise, consider renaming this test file.

File: ksql-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json [hunk 10]
```diff
@@ -226,14 +226,38 @@
...
+      "statements": [
+        "CREATE STREAM TEST (I INT, A ARRAY<ARRAY<INT>>) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "INSERT INTO TEST (I, A) VALUES (-1, ARRAY[ARRAY[1]]);"
+      ],
+      "inputs": [
+      ],
+      "outputs": [
+        {"topic": "test_topic", "key": null, "value": {"I": -1, "A": [[1]]}}
+      ]
+    },
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json]: I don't think we're fully testing the coercer with this test as the `1` from `ARRAY[ARRAY[1]]` will default to a `IntegerLiteral` and the type is `ARRAY<ARRAY<INT>>`.

What happens if you make it `ARRAY<ARRAY<BIGINT>>`?

Same goes for the map test below.

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: https://github.com/confluentinc/ksql/issues/4334

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: Fair enough

File: ksql-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json (hunk 10 above)
Comment (agavra) [reply ksql-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json]: done

**Pull request merged**